### PR TITLE
Retry concurrency exceptions in e2e/kubectl.go.  Fixes #26138

### LINKED ISSUE #26138: [k8s.io] Kubectl client [k8s.io] Kubectl rolling-update should support rolling-update to same image [Conformance] {Kubernetes e2e suite}
https://console.cloud.google.com/storage/kubernetes-jenkins/logs/kubernetes-e2e-gke/7634/

Failed: [k8s.io] Kubectl client [k8s.io] Kubectl rolling-update should support rolling-update to same image [Conformance] {Kubernetes e2e suite}

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:911
Expected error:
    <*errors.errorString | 0xc820684350>: {
        s: "Error running &{/workspace/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://104.154.237.36 --kubeconfig=/workspace/.kube/config rolling-update e2e-test-nginx-rc --update-period=1s --image=gcr.io/google_containers/nginx:1.7.9 --image-pull-policy=IfNotPresent --namespace=e2e-tests-kubectl-e1h8z] []  <nil> Error updating pod (Operation cannot be fulfilled on pods \"e2e-test-nginx-rc-zgrel\": the object has been modified; please apply your changes to the latest version and try again), retrying after 3 secondsError updating pod (Operation cannot be fulfilled on pods \"e2e-test-nginx-rc-zgrel\": the object has been modified; please apply your changes to the latest version and try again), retrying after 9 secondsError updating pod (Operation cannot be fulfilled on pods \"e2e-test-nginx-rc-zgrel\": the object has been modified; please apply your changes to the latest version and try again), retrying after 81 seconds Error from server: Operation cannot be fulfilled on pods \"e2e-test-nginx-rc-zgrel\": the object has been modified; please apply your changes to the latest version and try again\n [] <nil> 0xc82064c040 exit status 1 <nil> true [0xc8200e2348 0xc8200e2370 0xc8200e2398] [0xc8200e2348 0xc8200e2370 0xc8200e2398] [0xc8200e2358 0xc8200e2380] [0x9cb1a0 0x9cb1a0] 0xc820cab560}:\nCommand stdout:\nError updating pod (Operation cannot be fulfilled on pods \"e2e-test-nginx-rc-zgrel\": the object has been modified; please apply your changes to the latest version and try again), retrying after 3 secondsError updating pod (Operation cannot be fulfilled on pods \"e2e-test-nginx-rc-zgrel\": the object has been modified; please apply your changes to the latest version and try again), retrying after 9 secondsError updating pod (Operation cannot be fulfilled on pods \"e2e-test-nginx-rc-zgrel\": the object has been modified; please apply your changes to the latest version and try again), retrying after 81 seconds\nstderr:\nError from server: Operation cannot be fulfilled on pods \"e2e-test-nginx-rc-zgrel\": the object has been modified; please apply your changes to the latest version and try again\n\nerror:\nexit status 1\n",
    }
    Error running &{/workspace/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://104.154.237.36 --kubeconfig=/workspace/.kube/config rolling-update e2e-test-nginx-rc --update-period=1s --image=gcr.io/google_containers/nginx:1.7.9 --image-pull-policy=IfNotPresent --namespace=e2e-tests-kubectl-e1h8z] []  <nil> Error updating pod (Operation cannot be fulfilled on pods "e2e-test-nginx-rc-zgrel": the object has been modified; please apply your changes to the latest version and try again), retrying after 3 secondsError updating pod (Operation cannot be fulfilled on pods "e2e-test-nginx-rc-zgrel": the object has been modified; please apply your changes to the latest version and try again), retrying after 9 secondsError updating pod (Operation cannot be fulfilled on pods "e2e-test-nginx-rc-zgrel": the object has been modified; please apply your changes to the latest version and try again), retrying after 81 seconds Error from server: Operation cannot be fulfilled on pods "e2e-test-nginx-rc-zgrel": the object has been modified; please apply your changes to the latest version and try again
     [] <nil> 0xc82064c040 exit status 1 <nil> true [0xc8200e2348 0xc8200e2370 0xc8200e2398] [0xc8200e2348 0xc8200e2370 0xc8200e2398] [0xc8200e2358 0xc8200e2380] [0x9cb1a0 0x9cb1a0] 0xc820cab560}:
    Command stdout:
    Error updating pod (Operation cannot be fulfilled on pods "e2e-test-nginx-rc-zgrel": the object has been modified; please apply your changes to the latest version and try again), retrying after 3 secondsError updating pod (Operation cannot be fulfilled on pods "e2e-test-nginx-rc-zgrel": the object has been modified; please apply your changes to the latest version and try again), retrying after 9 secondsError updating pod (Operation cannot be fulfilled on pods "e2e-test-nginx-rc-zgrel": the object has been modified; please apply your changes to the latest version and try again), retrying after 81 seconds
    stderr:
    Error from server: Operation cannot be fulfilled on pods "e2e-test-nginx-rc-zgrel": the object has been modified; please apply your changes to the latest version and try again
    
    error:
    exit status 1
    
not to have occurred
```

### PR DESCRIPTION
**Pull request by pwittrock**
[Analytics]()

### CODE REVIEW

File: test/e2e/kubectl.go [hunk 1]
```diff
@@ -138,6 +138,28 @@ func readTestFileOrDie(file string) []byte {
 	return framework.ReadOrDie(path.Join(kubeCtlManifestPath, file))
 }
 
+const kubectlRetryStr = "the object has been modified; please apply your changes to the latest version and try again"
```
Comment (janetkuo) [inline test/e2e/kubectl.go]: This error is from `pkg/registry/generic/registry/store.go`. Consider `const` the error string there?

File: pkg/registry/generic/registry/store.go [hunk 2]
```diff
@@ -117,6 +117,8 @@ type Store struct {
 	Storage storage.Interface
 }
 
+const OptimisticLockErrorMsg = "the object has been modified; please apply your changes to the latest version and try again"
```
Comment (janetkuo) [inline pkg/registry/generic/registry/store.go]: use it in L320

File: test/e2e/kubectl.go [hunk 3]
```diff
@@ -138,6 +139,28 @@ func readTestFileOrDie(file string) []byte {
 	return framework.ReadOrDie(path.Join(kubeCtlManifestPath, file))
 }
 
+func runKubectlOrDieRetryOptimisticLock(data string, args ...string) string {
+	var err error
+	var output string
+	for i := 0; i < 2; i++ {
```
Comment (janetkuo) [inline test/e2e/kubectl.go]: Should we sleep and then retry?

File: test/e2e/kubectl.go [hunk 4]
```diff
@@ -907,7 +930,8 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
 			framework.WaitForRCToStabilize(c, ns, rcName, framework.PodStartTimeout)
 
 			By("rolling-update to same image controller")
-			framework.RunKubectlOrDie("rolling-update", rcName, "--update-period=1s", "--image="+nginxImage, "--image-pull-policy="+string(api.PullIfNotPresent), nsFlag)
+
+			runKubectlOrDieRetryOptimisticLock("rolling-update", rcName, "--update-period=1s", "--image="+nginxImage, "--image-pull-policy="+string(api.PullIfNotPresent), nsFlag)
```
Comment (janetkuo) [inline test/e2e/kubectl.go]: Suggest a more general name (we may want to retry on other errors in the future?), e.g. `runKubectlRetryOrDie`

File: test/e2e/kubectl.go [hunk 5]
```diff
@@ -138,6 +139,28 @@ func readTestFileOrDie(file string) []byte {
...
 }
 
+func runKubectlOrDieRetryOptimisticLock(data string, args ...string) string {
+	var err error
+	var output string
+	for i := 0; i < 2; i++ {
+		output, err = framework.RunKubectl(data, args)
+		if err == nil {
+			// All good - we are done
+			return output
```
Comment (janetkuo) [inline test/e2e/kubectl.go]: We still want `output` to be logged. Just `break` here.

Comment (janetkuo) [inline test/e2e/kubectl.go]: two `if`s can be combined

**Comment by pwittrock** [comment]
@k8s-bot node e2e test this issue: #26385

File: pkg/registry/generic/registry/store.go (hunk 2 above)
Comment (pwittrock) [inline pkg/registry/generic/registry/store.go]: Done

**Comment by janetkuo** [comment]
Just a reminder, some comments haven't been addressed.

File: test/e2e/kubectl.go (hunk 5 above)
Comment (pwittrock) [inline test/e2e/kubectl.go]: Good idea.

File: test/e2e/kubectl.go (hunk 4 above)
Comment (pwittrock) [inline test/e2e/kubectl.go]: Done

File: test/e2e/kubectl.go (hunk 3 above)
Comment (pwittrock) [inline test/e2e/kubectl.go]: Done

File: pkg/registry/generic/registry/store.go (hunk 2 above)
Comment (pwittrock) [inline pkg/registry/generic/registry/store.go]: Done

**Comment by pwittrock** [comment]
@janetkuo Thanks.  Pushing the updates now

**Comment by pwittrock** [comment]
PTAL

**Comment by janetkuo** [comment]
@k8s-bot e2e test this issue: #26210

**Comment by bprashanth** [comment]
I havent dug too deep into the failure logs so ignore me if this comment is totally off, but if a rolling update fails midway because of a conflict, is it safe to retry? or will we end up with 2 rcs that fight each other? maybe we should just modify whatever operation in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/rolling_updater.go that's conflicting, to retry instead? or did we push rolling update to deployment such that it's actually idempotent?

**Pull request merged**